src/lib.rs

```diff
@@ -1,41 +1,385 @@
 #![no_std]
 
-use gstd::{debug, msg, prelude::*};
+use gear_lib::non_fungible_token::token::{TokenId, TokenMetadata};
+use gstd::{async_main, exec, msg, prelude::*, ActorId};
+use nft_pixelboard_io::*;
 
-#[no_mangle]
-pub unsafe extern "C" fn handle() {
-    debug!("handle()");
-    let payload = String::from_utf8(msg::load_bytes()).expect("Invalid handle message");
+mod utils;
+use utils::*;
+
+fn get_pixel_count<P: Into<usize>>(width: P, height: P) -> usize {
+    let pixel_count = width.into() * height.into();
+    if pixel_count == 0 {
+        panic!("Width & height of a canvas/NFT must be more than 0");
+    };
+    pixel_count
+}
+
+fn check_painting(painting: &Vec<Color>, pixel_count: usize) {
+    if painting.len() != pixel_count {
+        panic!("`painting` length must equal a pixel count in a canvas/NFT");
+    }
+}
+
+fn check_pixel_price(pixel_price: u128) {
+    if pixel_price > MAX_PIXEL_PRICE {
+        panic!("`pixel_price` mustn't be more than `MAX_PIXEL_PRICE`");
+    }
+}
+
+fn get_mut_token_by_id<'a>(
+    rectangles: &'a BTreeMap<TokenId, Rectangle>,
+    tokens: &'a mut BTreeMap<Rectangle, TokenInfo>,
+    token_id: TokenId,
+) -> (&'a Rectangle, &'a mut TokenInfo) {
+    let rectangle = rectangles.get(&token_id).expect("NFT not found by an ID");
+    (
+        rectangle,
+        tokens
+            .get_mut(rectangle)
+            .expect("NFT not found by a rectangle"),
+    )
+}
+
+fn paint(
+    canvas_resolution: Resolution,
+    rectangle: &Rectangle,
+    rectangle_width: usize,
+    rectangle_height: usize,
+    canvas_painting: &mut [Color],
+    token_painting: Vec<Color>,
+) {
+    let canvas_width = canvas_resolution.width as usize;
+
+    let first_row_end = canvas_width * rectangle.upper_left_corner.y as usize
+        + rectangle.lower_right_corner.x as usize;
+    let first_row_start = first_row_end - rectangle_width;
```

Comment:
Should we check whether `first_row_end` is not less than `rectangle_width`? If no, please add a comment about the reason.

Reply:
No, because `rectangle.lower_right_corner.x`, that is added to `first_row_end`, is always greater than `rectangle_width`.

`rectangle_width` is calculated by substructing from `rectangle.lower_right_corner.x`:
https://github.com/gear-dapps/nft-pixelboard/blob/e9401078113c5e3d9e86a3d7d05ad4d7e823d115/io/src/lib.rs#L35-L37

And `self.lower_right_corner.x` is guaranteed to be greater than  `self.upper_left_corner.x` due to a check during minting here:
https://github.com/gear-dapps/nft-pixelboard/blob/e9401078113c5e3d9e86a3d7d05ad4d7e823d115/src/lib.rs#L103